server/src/main/java/org/apache/cassandra/sidecar/livemigration/LiveMigrationFileDownloader.java

```diff
@@ -0,0 +1,688 @@
+     * 2. Files with same name exists but size or last modified time doesn't match
+     *
+     * @param remoteFileAttrsMap map of remote instance file paths and their metadata (size, last modified time)
+     * @param localPath          file path on the local instance
+     * @param homeDir            file path of the data home directory
+     * @return true if the input path is a valid candidate for deletion, otherwise false
+     */
+    @VisibleForTesting
+    boolean canDelete(Map<String, FileAttributes> remoteFileAttrsMap,
+                      Path localPath,
+                      Path homeDir)
+    {
+        String absolutePath = localPath.toAbsolutePath().toString();
+        if (homeDir.equals(localPath))
+        {
+            return false;
+        }
+
+        // This condition handles both files and directories
+        if (!remoteFileAttrsMap.containsKey(absolutePath))
+        {
+            // Delete if the local file is not present at remote
+            return true;
+        }
+
+        try
+        {
+            FileProps fileProps = vertx.fileSystem().propsBlocking(absolutePath);
+
+            // We should not delete the entire directory based on timestamp because any file change can result in
+            // changing the timestamp of the directory.
+            if (!fileProps.isDirectory())
+            {
+                // The first condition includes a special case where time stamp changed but the file size did not change
+                // (eg: just flipped some bits etc.)
+                return fileProps.size() != remoteFileAttrsMap.get(absolutePath).size
+                       || (fileProps.lastModifiedTime() != remoteFileAttrsMap.get(absolutePath).lastModifiedTime);
+            }
+        }
+        catch (FileSystemException e)
+        {
+            LOGGER.error("{} Could not read properties for file={}", logPrefix, absolutePath, e);
+            throw e;
+        }
+        return false;
+    }
+
+    private void performDeletion(List<Path> filesToDelete)
+    {
+        AtomicInteger filesDeleted = new AtomicInteger(0);
+        AtomicInteger directoriesDeleted = new AtomicInteger(0);
+
+        for (Path localPath : filesToDelete)
+        {
+            if (!Files.exists(localPath))
+            {
+                continue;
+            }
+
+            if (Files.isDirectory(localPath))
+            {
+                LOGGER.debug("{} Deleting unwanted directory {}", logPrefix, localPath.toAbsolutePath());
+                vertx.fileSystem().deleteRecursiveBlocking(localPath.toAbsolutePath().toString(), true);
+
+                directoriesDeleted.incrementAndGet();
+            }
+            else
+            {
+                LOGGER.debug("{} Deleting unwanted file: {}", logPrefix, localPath.toAbsolutePath());
+                vertx.fileSystem().deleteBlocking(localPath.toString());
+                filesDeleted.incrementAndGet();
+            }
+        }
```

Comment:
When looping through a list of files to delete, can you catch any exception and continue with the next file? Assuming that you do not want to stop after a failure.